src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+            {
+                const int length = 16;
+                StringBuilder sb = StringBuilderCache.Acquire(length);
+                for (int i = 0; i < length; i++)
+                {
+                    byte[] randomNumber = new byte[1];
+                    rng.GetBytes(randomNumber);
+
+                    int rangeIndex = (randomNumber[0] % 3);
+                    rng.GetBytes(randomNumber);
+
+                    if (rangeIndex == 0)
+                    {
+                        // Get a random digit 0-9
+                        sb.Append((char)(AlphaNumChooser[rangeIndex] + randomNumber[0] % 10));
+                    }
+                    else
+                    {
+                        // Get a random alphabet in a-z, A-Z
+                        sb.Append((char)(AlphaNumChooser[rangeIndex] + randomNumber[0] % 26));
+                    }
+                }
+
+                return StringBuilderCache.GetStringAndRelease(sb);
+            }
+        }
+
+        private static string ComputeHash(string data, string algorithm)
+        {
+            if (algorithm == Sha256)
+            {
+                using (HashAlgorithm hash = SHA256.Create())
+                {
+                    return ComputeHash(data, hash);
+                }
+            }
+            else
+            {
+// Disable MD5 insecure warning.
+#pragma warning disable CA5351
+                using (HashAlgorithm hash = MD5.Create())
+                {
+                    return ComputeHash(data, hash);
+                }
+#pragma warning restore CA5351
+            }
+        }
+
+        private static string ComputeHash(string data, HashAlgorithm hash)
+        {
+            Encoding enc = Encoding.UTF8;
+            byte[] result = hash.ComputeHash(enc.GetBytes(data));
+
+            StringBuilder sb = StringBuilderCache.Acquire(result.Length * 2);
+            foreach (byte b in result)
+                sb.Append(b.ToString("x2"));
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        public class DigestResponse
+        {
+            public string Value => _value;
+            public Dictionary<string, string> Parameters => _parameters;
+
+            // Keep track of request values for this response.
+            public string Nonce;
+            public uint NonceCount;
+
+            private string _value;
+            private Dictionary<string, string> _parameters = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            public DigestResponse(string value)
+            {
+                Nonce = null;
+                NonceCount = 1;
+
+                _value = value;
+                Parse();
+            }
+
+            private unsafe string GetNextKey(ref char* p)
+            {
+                StringBuilder sb = StringBuilderCache.Acquire();
+
+                // Skip leading whitespace
+                while (*p == ' ')
+                {
+                    p++;
```

Comment:
It may not matter for perf in this case, but generally it's cheaper to change a local and then write back to the ref at the end rather than updating the ref on each operation.